plugins/soundsourcem4a/audiosourcem4a.cpp

```diff
@@ -0,0 +1,307 @@
+            if ((kMinSampleId + kSampleIdPrefetchCount) < sampleId) {
+                m_curSampleId = sampleId - kSampleIdPrefetchCount;
+            } else {
+                m_curSampleId = kMinSampleId;
+            }
+            m_curFrameIndex = (m_curSampleId - kMinSampleId) * kFramesPerSampleBlock;
+            // rryan 9/2009 -- the documentation is sketchy on this, but I think that
+            // it tells the decoder that you are seeking so it should flush its state
+            faacDecPostSeekReset(m_hDecoder, m_curSampleId);
+            // discard input buffer
+            m_inputBufferOffset = 0;
+            m_inputBufferLength = 0;
+        }
+        // decoding starts before the actual target position
+        DEBUG_ASSERT(m_curFrameIndex <= frameIndex);
+        const size_type prefetchFrameCount = frameIndex - m_curFrameIndex;
+        // prefetch (decode and discard) all samples up to the target position
+        DEBUG_ASSERT(frames2samples(prefetchFrameCount) <= m_prefetchSampleBuffer.size());
+        readFrameSamplesInterleaved(prefetchFrameCount, &m_prefetchSampleBuffer[0]);
+    }
+    DEBUG_ASSERT(m_curFrameIndex == frameIndex);
+    return frameIndex;
+}
+
+AudioSource::size_type AudioSourceM4A::readFrameSamplesInterleaved(
+        size_type frameCount, sample_type* sampleBuffer) {
+    sample_type* pSampleBuffer = sampleBuffer;
+    const diff_type readFrameIndex = m_curFrameIndex;
+    size_type readFrameCount = m_curFrameIndex - readFrameIndex;
+    while ((readFrameCount = m_curFrameIndex - readFrameIndex) < frameCount) {
+        if (isValidSampleId(m_curSampleId) && (0 == m_inputBufferLength)) {
+            // fill input buffer with block of samples
+            InputBuffer::value_type* pInputBuffer = &m_inputBuffer[0];
+            m_inputBufferOffset = 0;
+            m_inputBufferLength = m_inputBuffer.size(); // in/out parameter
+            if (!MP4ReadSample(m_hFile, m_trackId, m_curSampleId++,
+                    &pInputBuffer, &m_inputBufferLength,
+                    NULL, NULL, NULL, NULL)) {
+                qWarning() << "Failed to read MP4 sample data" << m_curSampleId;
+                break; // abort
+            }
+        }
+        if (0 == m_inputBufferLength) {
+            // EOF
+            break; // done
+        }
+        faacDecFrameInfo decFrameInfo;
+        decFrameInfo.bytesconsumed = 0;
+        decFrameInfo.samples = 0;
+        // decode samples into sampleBuffer
+        const size_type readFrameCount = m_curFrameIndex - readFrameIndex;
+        const size_type decodeBufferCapacityInBytes = frames2samples(
+                frameCount - readFrameCount) * sizeof(*sampleBuffer);
+        DEBUG_ASSERT(0 < decodeBufferCapacityInBytes);
+        void* pDecodeBuffer = pSampleBuffer;
+        NeAACDecDecode2(m_hDecoder, &decFrameInfo,
+                &m_inputBuffer[m_inputBufferOffset],
+                m_inputBufferLength / sizeof(m_inputBuffer[0]),
```

Comment:
m_inputBufferLength from the name it is not clear if it is in Byte or samples or frames. Can you find a better name?